internal/kubernetes/config.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+
+	k8s "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	httpconfig "github.com/open-telemetry/opentelemetry-collector-contrib/internal/http"
+)
+
+// AuthType describes the type of authentication to use for the K8s API
+type AuthType string
+
+const (
+	// AuthTypeNone means no auth is required
+	AuthTypeNone AuthType = "none"
+	// AuthTypeTLS means to use client TLS certs
+	AuthTypeTLS AuthType = "tls"
+	// AuthTypeServiceAccount means to use the built-in service account that
+	// K8s automatically provisions for each pod.
+	AuthTypeServiceAccount AuthType = "serviceAccount"
+	// AuthTypeKubeConfig uses local credentials like those used by kubectl.
+	AuthTypeKubeConfig AuthType = "kubeConfig"
+)
+
+var authTypes = map[AuthType]bool{
+	AuthTypeNone:           true,
+	AuthTypeTLS:            true,
+	AuthTypeServiceAccount: true,
+	AuthTypeKubeConfig:     true,
+}
+
+// K8sAPIConfig contains options relevant to connecting to the K8s API
+type K8sAPIConfig struct {
+	// How to authenticate to the K8s API server.  This can be one of `none`
+	// (for no auth), `tls` (to use manually specified TLS client certs, not
+	// recommended), `serviceAccount` (to use the standard service account
+	// token provided to the agent pod), or `kubeConfig` to use credentials
+	// from `~/.kube/config`.
+	AuthType AuthType `mapstructure:"auth_type"`
+	// TLS config options if `auth_type` is set to `tls`
+	TLSConfig httpconfig.TLSConfig `mapstructure:"tls_config"`
+}
+
+// Validate validates the K8s API config
+func (c *K8sAPIConfig) Validate() error {
+	if !authTypes[c.AuthType] {
+		return errors.New("Invalid authType for kubernetes: " + string(c.AuthType))
+	}
+
+	if c.AuthType == AuthTypeTLS && (c.TLSConfig.CertFile == "" || c.TLSConfig.KeyFile == "") {
+		return errors.New("for TLS auth, you must set both the kubernetesAPI.clientCertPath " +
+			"and kubernetesAPI.clientKeyPath config values")
+	}
+	return nil
+}
+
+// CreateRestConfig creates an Kubernetes API config from user configuration.
+func CreateRestConfig(apiConf *K8sAPIConfig) (*rest.Config, error) {
+	var authConf *rest.Config
+	var err error
+
+	if apiConf == nil {
+		authConf, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
Would return here and get rid of the else {} nesting since there's a lot of nesting below.